src/main/java/uk/gov/hmcts/reform/sscs/callback/handlers/ReciprocalLinkHandler.java

```diff
@@ -0,0 +1,107 @@
+import uk.gov.hmcts.reform.sscs.idam.IdamService;
+import uk.gov.hmcts.reform.sscs.idam.IdamTokens;
+
+@Slf4j
+@Service
+public class ReciprocalLinkHandler implements CallbackHandler<SscsCaseData> {
+
+    private final DispatchPriority dispatchPriority;
+
+    private final CcdService ccdService;
+
+    private final IdamService idamService;
+
+    @Autowired
+    public ReciprocalLinkHandler(CcdService ccdService,
+                                 IdamService idamService) {
+        this.dispatchPriority = DispatchPriority.LATEST;
+        this.ccdService = ccdService;
+        this.idamService = idamService;
+    }
+
+    @Override
+    public boolean canHandle(CallbackType callbackType, Callback<SscsCaseData> callback) {
+        requireNonNull(callback, "callback must not be null");
+        requireNonNull(callbackType, "callbacktype must not be null");
+
+        return callbackType.equals(CallbackType.SUBMITTED)
+            && (callback.getEvent() == EventType.VALID_APPEAL_CREATED
+            || callback.getEvent() == EventType.NON_COMPLIANT
+            || callback.getEvent() == EventType.INCOMPLETE_APPLICATION_RECEIVED);
+    }
+
+    @Override
+    public void handle(CallbackType callbackType, Callback<SscsCaseData> callback) {
+        if (!canHandle(callbackType, callback)) {
+            throw new IllegalStateException("Cannot handle callback");
+        }
+
+        log.info("Reciprocal link handler for case id {}", callback.getCaseDetails().getId());
+
+        IdamTokens idamTokens = idamService.getIdamTokens();
+
+        if (callback.getCaseDetails().getCaseData().getAppeal().getAppellant().getIdentity() != null
+            && callback.getCaseDetails().getCaseData().getAppeal().getAppellant().getIdentity().getNino() != null
+            && !callback.getCaseDetails().getCaseData().getAppeal().getAppellant().getIdentity().getNino().equals("")) {
+
+            String nino = callback.getCaseDetails().getCaseData().getAppeal().getAppellant().getIdentity().getNino();
+
+            List<SscsCaseDetails> matchedByNinoCases = getMatchedCases(nino, idamTokens);
+
+            if (matchedByNinoCases.size() > 0) {
+                log.info("Found " + matchedByNinoCases.size() + " matching cases for Nino " + callback.getCaseDetails().getCaseData().getGeneratedNino());
+
+                backLinkAssociatedCases(callback.getCaseDetails().getId(), matchedByNinoCases, idamTokens);
+            }
+        }
+    }
+
+    protected List<SscsCaseDetails> getMatchedCases(String nino, IdamTokens idamTokens) {
+        HashMap<String, String> map = new HashMap<String, String>();
+
+        map.put("case.appeal.appellant.identity.nino", nino);
+
+        return ccdService.findCaseBy(map, idamTokens);
+    }
+
+    private void backLinkAssociatedCases(Long caseId, List<SscsCaseDetails> matchedByNinoCases, IdamTokens idamTokens) {
+        if (matchedByNinoCases.size() > 0 && matchedByNinoCases.size() < 10) {
```

Comment:
is size limit 10 required?

Reply:
Yeh, they should only ever have 1 or 2 linked cases at most. So this is an upper limit to stop CCD getting hammered if someone tries to apply lots of times with different MRN dates.